common/transport/amqp/src/amqp.ts

```diff
@@ -135,6 +88,280 @@ export class Amqp {
+          '*': () => this._fsm.deferUntilTransition('connected')
+        },
+        connecting: {
+          _onEnter: (connectCallback) => {
+            let connectError = null;
+            const connectErrorHandler = (err) => {
+              connectError = err;
+            };
+
+            const amqpErrorHandler = (err) => {
+              debug('amqp10 client error: ' + err.toString());
+              this._fsm.handle('amqpError', err, () => {
+                if (this._disconnectHandler) {
+                  this._disconnectHandler(err);
+                }
+              });
+            };
+
+            // this._amqp.on('disconnected', () => {
+            //   this._fsm.transition('disconnecting', undefined, new errors.NotConnectedError());
+            // });
+
+            this._amqp.on(_amqpClientError, connectErrorHandler);
+            this._amqp.connect(this.uri)
+              .then((result) => {
+                debug('AMQP transport connected.');
+                this._amqp.on(_amqpClientError, amqpErrorHandler);
+                this._amqp.removeListener(_amqpClientError, connectErrorHandler);
+                this._fsm.transition('connected', connectCallback, result);
+                return null;
+              })
+              .catch((err) => {
+                this._amqp.removeListener(_amqpClientError, connectErrorHandler);
+                this._fsm.transition('disconnected');
+                /*Codes_SRS_NODE_COMMON_AMQP_16_003: [The `connect` method shall call the `done` callback if the connection fails.] */
+                this._safeCallback(connectCallback, connectError || err);
+              });
+          },
+          amqpError: (err, callback) => this._fsm.transition('disconnecting', callback, err),
+          connect: () => this._fsm.deferUntilTransition(),
+          disconnect: () => this._fsm.deferUntilTransition(),
+          '*': () => this._fsm.deferUntilTransition()
+        },
+        connected: {
+          _onEnter: (connectCallback, result) => {
+            /*Codes_SRS_NODE_COMMON_AMQP_16_002: [The `connect` method shall establish a connection with the IoT hub instance and if given as argument call the `done` callback with a null error object in the case of success and a `results.Connected` object.]*/
+            this._safeCallback(connectCallback, null, new results.Connected(result));
+          },
+          amqpError: (err, callback) => this._fsm.transition('disconnecting', callback, err),
+          connect: (callback) => callback(null, new results.Connected()),
+          disconnect: (disconnectCallback) => {
+            this._fsm.transition('disconnecting', disconnectCallback);
+          },
+          initializeCBS: (callback) => {
+            this._cbs = new ClaimsBasedSecurityAgent(this._amqp);
+            this._cbs.attach(callback);
+          },
+          putToken: (audience, token, callback) => {
+            this._cbs.putToken(audience, token, callback);
+          },
```

Comment:
Should there be a separate state for CBS initializing and CBS initialized?

Reply:
CBS state is handled with the CBS state machine. the connection itself doesn't depend on the CBS state so I don't think so. Plus, CBS isn't always used.

On the higher layer, we will separate disconnected/connecting/authenticating/authenticated/disconnecting

---
In reply to: [139511781](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139511781) [](ancestors = 139511781)